src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithCache.java

```diff
@@ -0,0 +1,98 @@
+package se.liu.ida.hefquin.engine.federation.access.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import se.liu.ida.hefquin.engine.federation.BRTPFServer;
+import se.liu.ida.hefquin.engine.federation.Neo4jServer;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.BRTPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.CardinalityResponse;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessException;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.federation.access.Neo4jRequest;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.StringResponse;
+import se.liu.ida.hefquin.engine.federation.access.TPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.TPFResponse;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+
+public class FederationAccessManagerWithCache implements FederationAccessManager 
+{
+	private FederationAccessManager fedAccMan;
+	private Map<TriplePattern, CompletableFuture<CardinalityResponse>> cacheMap;
+	
+	public FederationAccessManagerWithCache(FederationAccessManager fedAccMan) {
+		this.fedAccMan = fedAccMan;
+		this.cacheMap = new HashMap<TriplePattern, CompletableFuture<CardinalityResponse>>();
+	}
+	
+	protected void addResultToCache(TriplePattern tp, CompletableFuture<CardinalityResponse> response) {
```

Comment:
```suggestion
	protected void addResultToCache(final TriplePattern tp, final CompletableFuture<CardinalityResponse> response) {
```

Reply:
Additionally, I prefer helper methods such as this one to be placed at the end of the file, after the public methods.